week-2/mandatory/3-api/server.js

```diff
@@ -0,0 +1,382 @@
+    res.status(400).json("The price needs to be larger than 0");
+  }
+});
+
+app.post("/customers/:customerId/orders", (req, res) => {
+  let id = parseInt(req.params.customerId),
+    sql_orders =
+      "insert into orders (order_date, order_reference, customer_id)" +
+      "values (current_date,$1,$2) returning order_reference",
+    sql_customers = "select * from customers where id = $1",
+    sql_orderNumber =
+      "select order_reference from orders order by id DESC limit 1";
+
+  db.query(sql_customers, [id], (error, result) => {
+    if (error) {
+      console.log("pg error code:", error.code);
+      res.status(500).json(serverError);
+    } else if (result.rows.length > 0) {
+      db.query(sql_orderNumber, (error, result) => {
+        if (error) {
+          console.log("pg error code:", error.code);
+          res.status(500).json(serverError);
+        } else {
+          let lastOrder = result.rows[0].order_reference;
+          //orderRef functions are borrowed from https://electrictoolbox.com/pad-number-zeroes-javascript/ and SOF site
+          let orderRef = lastOrder.replace(/(\d+)+/g, function (match, number) {
+            function pad(number, length) {
+              let str = "" + number;
+              while (str.length < length) {
+                str = "0" + str;
+              }
+              return str;
+            }
+            return pad(parseInt(number) + 1, 3);
+          });
+          db.query(sql_orders, [orderRef, id], (error, result) => {
+            if (error) {
+              console.log("pg error code:", error.code);
+              res.status(500).json(serverError);
+            } else {
+              res
+                .status(201)
+                .json(
+                  `Order placed with reference ${result.rows[0].order_reference}`
+                );
+            }
+          });
+        }
+      });
+    }
+  });
+});
+
+app.put("/customers/:customerId", (req, res) => {
+  const id = parseInt(req.params.customerId),
+    name = req.body.name,
+    address = req.body.address,
+    city = req.body.city,
+    country = req.body.country;
+  db.query("SELECT * FROM customers where id = $1", [id], (error, result) => {
+    if (error) {
+      console.log("pg error code:", error.code);
+      res.status(500).json(serverError);
+      throw error;
+    } else {
+      if (result.rows.length > 0) {
+        db.query(
+          "update customers set name = $1, address = $2, city = $3,country = $4 where id = $5",
+          [name, address, city, country, id],
+          (error) => {
+            if (error) {
+              console.log("pg error code:", error.code);
+              res.status(500).json(serverError);
+              throw error;
+            } else {
+              res.status(200).json(`Customer with ID ${id} data updated.`);
+            }
+          }
+        );
+      } else {
+        res.status(404).json(`There is no customer with this ID ${id}`);
+      }
+    }
+  });
+});
```

Comment:
Good use of different http status code here, to distinguish between user error and server error.